### PR TITLE
RSDK-2895 Add jetson support for csi module

### PR DESCRIPTION
**Pull request by seanavery**
### Description

A [viam module](https://docs.viam.com/program/extend/modular-resources/) for CSI cams! It is written in cpp via the [viam-cpp-sdk](https://github.com/viamrobotics/viam-cpp-sdk) to create a [camera component](https://docs.viam.com/components/camera/webcam/) . The `get_image` api is fed from a Gstreamer pipeline utilizing the hardware-accelerated ISP functionality in the `nvarguscamerasrc` plugin.

```
.
├── CMakeLists.txt
├── Makefile
├── README.md
├── constraints.h
├── csi_camera.h
├── csi_camera.cpp
├── main.cpp
├── doc
│   └── SUPPORT.md
├── etc
│   ├── Dockerfile.jetson
│   ├── Dockerfile.test.jetson
│   ├── app-config.json
│   ├── viam-csi-jetson-arm64.yml
│   ├── viam-logo.jpeg
│   └── viam-server.png
├── test
│   ├── perf_sdk.cpp
│   ├── test_csi_camera.cpp
│   ├── test_jetson_pipeline.sh
```

`main.cpp` Entrypoint to register module and start grpc server.
`csi_camera.cpp` Implements Camera interface and GStreamer functionality.
`constraints.h` Default params.

### Todo 
- Currently, the module is packaged as an appimage using appimage-builder pipeline. A pure static build would be preferable but still need to get that working.
- ~~Github CI workflow not working right now due to slow build times.~~
- ~~Have not yet made a concerted effort to reduce the size of the appimage bundle.~~
- Add perf script for sdk throughput.
- Need to swap out exits for throws and add error handling to camera server in the cpp-sdk.
- ~~Upload image and change install instructions to point to `packages.viam.com`.~~
- Add `aix` update mechanism to the appimage.

### CODE REVIEW

File: csi_camera.cpp
Comment (tessavitabile) [inline csi_camera.cpp:251]: Sleep so we don't busy loop?

File: etc/app-config.json
Comment (tessavitabile) [inline etc/app-config.json:11]: What if there are multiple cameras plugged in? How will the resource know which one to connect to?

File: csi_camera.cpp
Comment (tessavitabile) [inline csi_camera.cpp:207]: If this causes the process to exit, then that will kill the module process, which is a problem if there are multiple instances running within the module. Instead, just this instance should behave as closed. Same for the other EXITs.

File: csi_camera.cpp
Comment (tessavitabile) [inline csi_camera.cpp:451]: Does a modular resource not have to implement close()?

**Review by tessavitabile — COMMENTED** [review]
Nice work! I added a few comments, but you don't need to wait for my sign-off before merging.

File: etc/app-config.json
Comment (tessavitabile) [reply etc/app-config.json:11]: Ah, I see we do let you specify video_path. Can you add this to the sample config?

File: csi_camera.cpp
Comment (seanavery) [reply csi_camera.cpp:437]: There is no explicit "close" api AFAIK.

From docs:

- During shutdown, a RemoveResource() is sent for every resource handled by a module, instructing the module to close and remove it.
- Lastly, the module process itself is stopped by the parent, before the parent itself terminates.

Added deconstructor to make sure GST pipeline closes down when the resource is removed.

File: csi_camera.cpp
Comment (seanavery) [reply csi_camera.cpp]: I added a log if no bytes are received.

Not sure what else can be done here since it is possible that an image is simple not available and there is nothing wrong with the pipeline.

The `read` method in RDK camera client will handle the empty frame by erroring during the decode step.

File: etc/Dockerfile.test.jetson
Comment (seanavery) [reply etc/Dockerfile.test.jetson:11]: Yes - removed stale files.

File: README.md
Comment (seanavery) [reply README.md:7]: Ok sounds good.

I added a warning since the underlying viam-cpp-sdk is still in alpha.

https://github.com/viamrobotics/viam-cpp-sdk#instability-notice

Comment (tessavitabile) [reply README.md:7]: Ah, sounds good

File: csi_camera.cpp
Comment (tessavitabile) [reply csi_camera.cpp]: Ok cool, sounds like there's nothing to do

File: csi_camera.cpp
Comment (tessavitabile) [reply csi_camera.cpp:438]: Ok cool, if resources support close by implementing a destructor, then that sounds fine

File: csi_camera.cpp
Comment (bazile-clyde) [inline csi_camera.cpp:119]: Nit: no need for this since we have the `override` specifier for the methods that need them.

File: csi_camera.cpp
Comment (bazile-clyde) [inline csi_camera.cpp:131]: We'd only need to reconfigure if the resolution or frame format changes, no? Is that checked somewhere? I'm not 100% sure when this function will get called, but we wouldn't want to reconfigure this camera anytime _anything_ in the config changes.

File: csi_camera.cpp
Comment (bazile-clyde) [inline csi_camera.cpp:121]: This duplicates the logic in the destructor and then the constructor. It is better to put the logic for both functions into private functions and call those private functions where needed, that is, in the constructor, destructor, and reconfigure functions. That way, we won't later modify code in one place and forget to do it in another.

**Review by bazile-clyde — CHANGES_REQUESTED** [review]

File: csi_camera.cpp
Comment (seanavery) [reply csi_camera.cpp:124]: Good call. Abstracted into `init` fn.

File: csi_camera.cpp
Comment (seanavery) [reply csi_camera.cpp:22]: Yes, there is only one frame format available `'RG10' (10-bit Bayer RGRG/GBGB)` 
This is the raw bayer pattern.

However, when pulling from `nvarguscamerasrc` we just specify `video/x-raw(memory:NVMM)`  which uses DMA buffer with pixel format NV12" (YUV 4:2:0). 

So, `format` should be simply ignored. Will need to write documentation on the attribute options.

File: csi_camera.cpp
Comment (seanavery) [reply csi_camera.cpp:126]: Good point. 

RDK will handle only sending `reconfigure` calls when the module or associated component params are changed.

If a change is made to the following params, GST changes are not required.
```
"name": "csicam_test",
"namespace": "rdk",
"type": "camera",
"debug": true,
"model": "viam:camera:csi"
```

File: csi_camera.cpp
Comment (seanavery) [reply csi_camera.cpp:228]: Yes this loop simply hangs execution until state change occurs in the GST pipeline. 

I did not want to do this in the background as it is much cleaner to handle errors/exits synchronously.

File: README.md
Comment (seanavery) [reply README.md:27]: Yes, ideally I can change install instructions to a `Module Registry` link : )

File: csi_camera.cpp
Comment (tessavitabile) [reply csi_camera.cpp:228]: We don't need to do this in the background, but we should still sleep in the body of this while loop. We should also add a timeout, so we don't wait indefinitely.

File: main.cpp
Comment (seanavery) [reply main.cpp:11]: Yeah... I was only using it in one location but probably better to make header file with pragma once for good measure.

Comment (seanavery) [reply main.cpp]: ok, now updated with header file pattern.

File: csi_camera.cpp
Comment (seanavery) [reply csi_camera.cpp]: Added const for appropriate variables.

File: CMakeLists.txt
Comment (seanavery) [reply CMakeLists.txt:4]: This is the version I was on from personal repo - can roll back to 0.0.1.

File: csi_camera.cpp
Comment (seanavery) [reply csi_camera.cpp]: Great point! Added 1 second timeout.

File: etc/app-config.json
Comment (seanavery) [reply etc/app-config.json:11]: Yep, added to sample config.

File: csi_camera.cpp
Comment (seanavery) [reply csi_camera.cpp]: Removed device fetching. Will add back in for Pi support.

File: csi_camera.cpp
Comment (seanavery) [reply csi_camera.cpp]: thanks! fixed.

File: csi_camera.cpp
Comment (seanavery) [reply csi_camera.cpp]: Looks like libargus will break with anything less than 100 pixels. 

Anything reasonable should work here so not sure if it necessary to add extra checks for size.

File: etc/viam-csi-jetson-arm64.yml
Comment (seanavery) [reply etc/viam-csi-jetson-arm64.yml:1]: hi @Otterverse . Curious to see if you have any input on appimage builder.

The build is quite large (130 MB) - I attempted to slim down by only including gstreamer pugin groups I am using [base, good].

Also, looking for some input on how to handle updates.

Comment (Otterverse) [reply etc/viam-csi-jetson-arm64.yml:1]: To reduce size, you may want to add an "exclude" section to the files section near the bottom. Typically I exclude all of /usr/share/doc and /usr/share/man to start with (no need for documentation and man pages inside an appimage.) You can probably find a lot of other stuff as well.

If you run the built appimage with --appimage-extract it'll extract the contents to a folder called "squashfs-root" and you can browse through that for extra cruft to exclude.

**Review by Otterverse — COMMENTED** [review]
I'm only really looking over the appimage packaging parts of this, as requested. Feel free to message on slack if you'd like to discuss anything further. I'm happy to explain in more detail my usual approach to appimages and/or share whatever knowledge I have there.

File: csi_camera.cpp
Comment (seanavery) [reply csi_camera.cpp]: Added table JETSON.md with table of attribute options.

File: etc/viam-csi-jetson-arm64.yml
Comment (seanavery) [reply etc/viam-csi-jetson-arm64.yml]: Thanks!

Yes, realized that I was in fact not bundling GStreamer. 

I was able to completely remove `apt` section and only explicitly include `/usr/local/lib/libviamsdk.so.noabi` which automatically bundles all dependencies of the sdk.

The necessary GStreamer libraries come with standard `jetpack` on Jetson devices, so no need to bundle in the image.

The size of the libviamsdk.so and dependencies is still 130 MB - not sure if it is possible to shrink more.

File: csi_camera.cpp
Comment (seanavery) [reply csi_camera.cpp]: ok I see what you saying.

Added 5 second timeout to busy loop. Also waiting 10 ms between state queries to avoid using up too much cpu.

File: csi_camera.cpp
Comment (seanavery) [reply csi_camera.cpp]: Interesting, can you explain a little more wrt multiple instances? I thought I wanted to exit the entire process.

Comment (Otterverse) [reply csi_camera.cpp]: So any specific module supports one or more particular models of hardware. Here, it supports this specific type of camera. But if someone has two (or more) of these cameras on the same robot, then they'll both be using this same module. In such a case, there would be multiple configured resources in the robot, with different names, but each would be the same model, and would be supported by the same module. AddResource() would get called twice, once for each instance. Closing or removing one camera from the config should never interfere with the other camera. And even if ALL cameras are removed, the module itself should stay running (though pretty much in an idle state) until the module itself is Close()ed as part of robot shutdown, as it's possible someone may add a camera back if they're adjusting configs, and the module itself needs to remaining listening for another AddResource() call in case that happens.

File: csi_camera.cpp
Comment (tessavitabile) [reply csi_camera.cpp]: Thank you!

File: csi_camera.cpp
Comment (seanavery) [reply csi_camera.cpp]: Ok this is all making sense - if simply exit it will kill all of the instantiations. 

For example, [here](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/sdk/services/mlmodel/server.cpp#L114C19-L118C24) the ml model service handles exceptions and signals over grpc.

I need to add this status wrapper to all the methods in [components/camera/server.cpp](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/sdk/components/camera/server.cpp) to handle throws and send a failed status.

Comment (seanavery) [reply csi_camera.cpp]: Since this requires an sdk change, I am wondering if i should crete multi-cam support ticket and save for subsequent PR.

Comment (tessavitabile) [reply csi_camera.cpp]: I think that makes sense.

File: README.md
Comment (seanavery) [reply README.md:12]: Ok - working on script to upload to `packages.viam.com`. Will update the install instructions to point there.

File: csi_camera.cpp
Comment (seanavery) [reply csi_camera.cpp]: Ok, upgraded to c++ 17 and now have very concise attribute parsing.

File: Makefile
Comment (Otterverse) [inline Makefile:137]: Just saw this push and wanted to make a note. We use ghcr.io (the github container registry) for our other images, like canon. I think that would be preferable to use here as well, rather than an individual user's account on dockerhub, as it more easily integrates with our github permissions and teams.

File: main.cpp
Comment (bazile-clyde) [inline main.cpp:23]: nit: typo in "regsitration"

File: csi_camera.h
Comment (bazile-clyde) [inline csi_camera.h:63]: nit: do these methods need to be public?

File: csi_camera.cpp
Comment (bazile-clyde) [inline csi_camera.cpp:72]: Why are we not setting this to the given `mime_type`? It looks like that argument is ignored. If so, the debug log is misleading since we aren't expecting the MIME type given in the argument but always the default one.

File: csi_camera.cpp
Comment (bazile-clyde) [inline csi_camera.cpp:56]: Nice.

File: csi_camera.cpp
Comment (bazile-clyde) [inline csi_camera.cpp:1]: The comments and logs throughout this file are helpful! They make the code a lot more readable and easier to reason about.

File: constraints.h
Comment (bazile-clyde) [inline constraints.h:15]: How to do we set this to `true` to use the `FAKE_CAMERA` functionality?

File: CMakeLists.txt
Comment (bazile-clyde) [reply CMakeLists.txt]: Was just curious. Either is fine by me.

File: main.cpp
Comment (seanavery) [reply main.cpp:71]: Yes I can remove this - just added default socket for testing purposes.

File: main.cpp
Comment (seanavery) [reply main.cpp:23]: thanks!

File: main.cpp
Comment (seanavery) [reply main.cpp:21]: I do not think it makes sense to define handlers using std::signal since the "handler" is `server->shutdown()` in the sdk.

We simply block until signal is received then call shutdown in the sdk.

File: constraints.h
Comment (seanavery) [reply constraints.h:15]: Great point - added `fake` attribute to config so you can set in app.

File: test/test_csi_camera.cpp
Comment (seanavery) [reply test/test_csi_camera.cpp:64]: Added checks.

File: test/test_jetson_pipeline.sh
Comment (seanavery) [reply test/test_jetson_pipeline.sh:1]: This is a shell script that runs GST pipeline using `gst-launch-1.0`. Basically just an easy way to check that camera/gst is working. If it is not helpful I can remove.

File: csi_camera.cpp
Comment (seanavery) [reply csi_camera.cpp:73]: Currently do not have a way to change mime type based on server request. So we always set the mimetype to default (jpeg).

File: test/perf_sdk.cpp
Comment (seanavery) [reply test/perf_sdk.cpp:1]: This is for `fake` camera file fetch.

File: csi_camera.h
Comment (seanavery) [reply csi_camera.h:64]: Not right now. I was leaving these public for unit testing later on.

File: main.cpp
Comment (bazile-clyde) [reply main.cpp]: >I do not think it makes sense to define handlers using std::signal since the "handler" is server->shutdown() in the sdk

I see. Will resolve.

File: test/test_jetson_pipeline.sh
Comment (bazile-clyde) [reply test/test_jetson_pipeline.sh:1]: Ah I see. It does seem helpful. You can document it on one of the markdown files so people know what it's here for.

File: main.cpp
Comment (bazile-clyde) [reply main.cpp:75]: I'm not seeing any tests that use this. If this and `fake` are for future use I'd just leave them out for now. The test creator could modify this as needed and Ideally would just stub out the functions they need to. But I'm not going to block the code review on this so it's completely up to you.

File: constraints.h
Comment (bazile-clyde) [reply constraints.h:15]: If it's for internal testing purposes we probably don't want to document that and confuse users. Or see my other comment about leaving it out of this PR.

**Review by bazile-clyde — APPROVED** [review]
Left a few optional comments but LGTM.

File: constraints.h
Comment (seanavery) [reply constraints.h:15]: Removed fake camera.

**Pull request merged**
